articles/qos_configurability.md

```diff
@@ -0,0 +1,374 @@
+
+### Parameter callbacks
+
+If we want the following example to work:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  [] (const QoSProfile & qos) -> bool{
+      return qos.is_reliable();
+  });
+```
+
+We could completely ignore parameters [on set callbacks](https://github.com/ros2/rclcpp/blob/3defa8fc9d7410bd833ecd95b305ac94bb9b627a/rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp#L180-L192), the publisher creation code would do something like:
+
+```cpp
+rcl_interfaces::msg::ParameterDescription policy_x_description{};
+policy_x_description.read_only = true;
+policy_x_description.description = "<string with a description of the policy, including node and topic name>";
+auto policy_x_value = node->declare_parameter(
+    "<name of the parameter>", qos_provided_in_code.<getter_for_policy_x>(), policy_x_description);
+// Repeat for all policies
+rclcpp::QoS final_qos();
+// construct `final_qos` from the parameter values read above
+if (!user_provided_callback(final_qos) {
+    // qos profile rejected, throw error here
+}
+```
+
+In that way, we avoid installing a parameter callback that will be triggered when any parameter is modified.
+
+Note: this seems to be what you generally want in the case of read-only parameters.
+Note 2: We could also have a `declare_parameters_atomically` method and use:
+  ```cpp
+  auto callback_handle = node->add_on_set_parameters_callback(wrapper_for_qos_callback(user_provided_callback));
+  node->declare_parameters_atomically(...);  // names and defaults for all parameters given here
+  // callback_handle goes out of scope, the callback is auto-removed.
+  ```
+  `declare_parameters_atomically` would be needed, if not you cannot have conditions like `history == keep_all or history_depth > 10`.
+
+### Parameters URIs
+
+Parameters URIs currently look like:
+
+```
+rosparam://my/full/namespace/node_name/param_group.nested_param_group.param_name
+```
+
+For the following QoS policy:
+- Full node name: `/asd/bsd/my_node`
+- Entity: `publisher`
+- Topic name: `/foo/bar/topic_name`
+- Policy: `reliability`
+
+The URI of the automatically generated parameter would be:
+
+```
+rosparam://asd/bsd/my_node/qos_profiles.publisher./foo/bar/topic_name.reliability
+```
+
+The `/` in one of the subgroups looks odd, but it's currently accepted.
+
+If in the future the addressing changes described in [#241](https://github.com/ros2/design/pull/241) are implemented, the URI would be:
+
+```
+rosparam://asd.bsd.my_node/qos_profiles/publisher/foo/bar/topic_name/reliability
+```
+
+but in both cases the previous parameters file example wouldn't change.
+
+## Caveats
+
+### Creating more than one publisher/subscription in the same topic with different QoS
+
+If only the topic name is used for the parameter name, it would be impossible to create two publishers with different QoS in the same topic.
+
+That could be solved by adding an optional extra identifier:
```

Comment:
Can two separate publishers, using the same topic/message_type combination use different QoS settings? If this is a valid use case, I can understand also introducing an ID to further discriminate. Personally, I would prefer a scheme where topic/message_type was required and ID was a further option.

Reply:
> Can two separate publishers, using the same topic/message_type combination use different QoS settings?

Yes, they can.
I'm not sure if that has an use case or not.

Reply:
@wjwwood @jacobperron any opinions on this?
Should it be possible to reconfigure the QoS of different publishers in the same topic (name and type) to different QoS within a node?
Should it be assumed that all publishers in a topic within a node use the same QoS?

Reply:
I assume the case is rare, but I think it should be possible. E.g. by default one publisher may be using *transient local* durability and another using *volatile*. Perhaps the node author would like the history depth of each publisher to be externally configured independently. Having the optional identifier makes this possible.

Reply:
I think a reasonable use case for this would be to have two subscriptions in a node, each with the same settings except for the history depth. Imagine two libraries created subscriptions to the same topic (and same type) using the same node (which was passed into the libraries). The two libraries are not coordinating and so may want different QoS. They may also do this to just have two different history caches (so one library taking a sample doesn't interfere with the other library).

So having two entities with the same name and type but different QoS is a valid use case from my perspective.